### PR TITLE
feat(iam): Permissions Boundaries

### LINKED ISSUE #3242: Applying permissions boundary to aws-cdk globally
**Note: for support questions, please first reference our [documentation](https://docs.aws.amazon.com/cdk/api/latest), then use [Stackoverflow](https://stackoverflow.com/questions/ask?tags=aws-cdk)**. This repository's issues are intended for feature requests and bug reports.

* **I'm submitting a ...**
  - [ ] :beetle: bug report
  - [x] :rocket: feature request
  - [ ] :books: construct library gap
  - [ ] :phone: security issue or vulnerability => Please see [policy](https://github.com/awslabs/aws-cdk/security/policy)
  - [ ] :question: support request => Please see note at the top of this template.

* **What is the current behavior?**
 It’s not possible to set permissions boundary globally for cdk, or anything similar along those lines

* **What is the expected behavior (or behavior of feature suggested)?**

My Gitlab CI/CD pipeline runner uses an IAM role that has permissions boundary set (only what our devops team is willing to let us do on our own). However, cdk tries to automagically create or alter roles/permissions/security groups/etc between resources that don’t fall under rules set in our permissions boundary policy (i.e. any customer role must have /custom/ path, etc) . We end up with bunch of errors (cdk trying to create something that permissions boundary on IAM role for the gitlab runner doesn’t allow). What I’m asking is… it would be great if there a way to tell cdk to somehow use permissions boundary policy whenever it tries to automagically create relationships (IAM roles, policies, etc) between the resources.

* **What is the motivation / use case for changing the behavior or adding this feature?**

We would be able to limit cdk based on pre-created permission boundaries created by our ops team.

* **Please tell us about your environment:**
  
  - CDK CLI Version: 0.37.0
  - Module Version: 0.37.0
  - OS: Mac
  - Language: TypeScript

* **Other information** (e.g. detailed explanation, stacktraces, related issues, suggestions how to fix, links for us to have context, eg. associated pull-request, stackoverflow, gitter, etc)

Conversation w/ @eladb on Gitter.

image

### PR DESCRIPTION
**Pull request by rix0rrr**
Allow configuring Permissions Boundaries for an entire subtree using
Aspects, add a sample policy which can be used to reduce future
misconfiguration risk for untrusted CodeBuild projects as an example.

Addresses one part of aws/aws-cdk-rfcs#5.

Fixes #3242.

ALSO IN THIS COMMIT:

Fix a bug in the `assert` library, where `haveResource()` would *never* match
any resource that didn't have a `Properties` block (even if we tested for no property
in particular, or the absence of properties). This fix caused two ECS tests to fail,
which were asserting the wrong thing anyway (both were asserting `notTo(haveResource(...))`
where they actually meant to assert `to(haveResource())`.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by robertd** [comment]
@rix0rrr  https://media2.giphy.com/media/3o7aCWJavAgtBzLWrS/source.gif

File: packages/@aws-cdk/aws-codebuild/lib/untrusted-code-boundary-policy.ts [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+            'ec2:CreateNetworkInterfacePermission',
+            'ec2:CreateNetworkInterface',
+            'ec2:DescribeNetworkInterfaces',
+            'ec2:DeleteNetworkInterface',
+            'ec2:DescribeSubnets',
+            'ec2:DescribeSecurityGroups',
+            'ec2:DescribeDhcpOptions',
+            'ec2:DescribeVpcs',
+
+            // NOTABLY MISSING:
```
Comment (eladb) [inline packages/@aws-cdk/aws-codebuild/lib/untrusted-code-boundary-policy.ts:76]: Would it make sense to add a "forbid" statements for these?

File: packages/@aws-cdk/aws-iam/README.md [hunk 2]
```diff
@@ -264,6 +264,39 @@ const newPolicy = new Policy(stack, 'MyNewPolicy', {
...
+the Identity Policy and the Permissions Boundary attached to the Role (or
+User). Permissions Boundaries are typically created by account
+Administrators, and their use on newly created `Role`s will be enforced by
+IAM policies.
+
+It is possible to attach Permissions Boundaries to all Roles created in a construct
+tree all at once:
+
+```ts
+// This imports an existing policy. You can also create a new one.
```
Comment (eladb) [inline packages/@aws-cdk/aws-iam/README.md:284]: Don't we also want to show how to define a new permission boundary using the CDK?

**Review by eladb — APPROVED** [review]
Looks awesome! People are going to love this!

Add a note in the PR description that this is related to https://github.com/aws/aws-cdk-rfcs/issues/5

**Comment by eladb** [comment]
Minor notes, added `do-not-merge`

File: packages/@aws-cdk/aws-codebuild/lib/untrusted-code-boundary-policy.ts (hunk 1 above)
Comment (rix0rrr) [reply packages/@aws-cdk/aws-codebuild/lib/untrusted-code-boundary-policy.ts:76]: Not really. It does the same thing.

File: packages/@aws-cdk/aws-iam/README.md (hunk 2 above)
Comment (rix0rrr) [reply packages/@aws-cdk/aws-iam/README.md:284]: The 99.9% use case is that a PB will have been created for you by an administrator. It only makes sense in very select cases to create one yourself. But sure.

**Pull request merged**